Mypy checking for build.py

Volker-Weissmann opened the pull request:
`run_mypy.py` currently does not check `mesonbuild/build.py`, much to my distaste. Currently, mypy would show a lot of warnings in build.py if it did.

I fixed all but one mypy warning in build.py.
Some of these mypy warnings hinted at real bugs, I fixed those.
The 3 test cases I added were inspired by the mypy warnings and my experiences fixing them.

There is still one mypy warning left:
```
mesonbuild/build.py:2125:40: error: Argument 2 to "join" has incompatible type "Union[str, bool]"; expected "str"  [arg-type]
```
That one hints at a real bug that I can use to create a crash. See #11748.

What do you think? Should I do `ninjabackend.py` next? I suspect that #1633 is hiding in the mypy warnings of ninjabackend.py


And no, I'm not gaming any LOC metric here. :-)

Volker-Weissmann committed:
Fix error message when an unknown generated file is passed to `build_tgt.extract_objects`

Volker-Weissmann committed:
Fix potential `get_transitive_link_deps_mapping` caching bug

`get_transitive_link_deps_mapping` has an `@lru_cache`. Therefore, we should
never modify its output, only copies of its output.

Volker-Weissmann committed:
Minor cleanup: Remove `get_java_args`

eli-schwartz commented on mesonbuild/build.py:
```diff
@@ -671,7 +671,7 @@ def parse_overrides(kwargs: T.Dict[str, T.Any]) -> T.Dict[OptionKey, str]:
         # provided by typed_kwargs
         if isinstance(opts, dict):
-            return T.cast('T.Dict[OptionKey, str]', opts)
+            return T.cast(T.Dict[OptionKey, str], opts)
```
same.

Volker-Weissmann replied:
Fixed

eli-schwartz commented on mesonbuild/build.py:
```diff
@@ -706,22 +707,22 @@ def __init__(
             sources: T.List['SourceOutputs'],
             structured_sources: T.Optional[StructuredSources],
-            objects: T.List[ObjectTypes],
+            objects: T.Sequence[T.Union[str, File, 'ExtractedObjects']],
```
I would think we need a new type alias for this. Also, list > sequence.

eli-schwartz reviewed (commented):
Only took a quick look, here are my initial thoughts.

Volker-Weissmann commented on mesonbuild/modules/i18n.py:
```diff
@@ -62,9 +60,7 @@ class Gettext(TypedDict):
-            str, build.BuildTarget, build.CustomTarget, build.CustomTargetIndex,
-            build.ExtractedObjects, build.GeneratedList, ExternalProgram,
-            mesonlib.File]]
+            str, mesonlib.File, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList]]
```
Passing `ExternalProgram` or `ExtractedObjects` results in an unhandled exception before this PR.

As to why I also removed build.BuildTarget:
The argument will be passed to msgfmt or itstool eventually.

These tools accept *.mo, *.xml and *.desktop files. (They don't check the filename, but just care about the content.) I don't see a way how a BuildTarget could also be a *valid and useful* *.mo or *.xml or *.desktop file.

If you pass an ELF file to msgfmt, it will print warnings, and output an empty file.

Volker-Weissmann commented on mesonbuild/backend/backends.py:
```diff
@@ -1642,15 +1642,18 @@ def generate_target_install(self, d: InstallData) -> None:
             assert len(t.install_tag) == num_out
+            for x in outdirs_unchecked:
+                assert not x is True
+            outdirs = T.cast(T.List[T.Union[str, T.Literal[False]]], outdirs_unchecked)
```
Yes, thats what made the tests fail

Volker-Weissmann commented on mesonbuild/backend/backends.py:
```diff
@@ -1233,14 +1233,14 @@ def create_test_serialisation(self, tests: T.List['Test']) -> T.List[TestSeriali
         pickle.dump(self.create_test_serialisation(tests), datafile)
 
-    def construct_target_rel_paths(self, t: T.Union[build.Target, build.CustomTargetIndex], workdir: T.Optional[str]) -> T.List[str]:
+    def construct_target_rel_paths(self, t: T.Union[build.Target, build.CustomTargetIndex], workdir: T.Optional[str]) -> T.Sequence[str]:
```
> `T.Sequence[str]` will accept that and also `'foo'`

I did not know that.

> What exactly is wrong with it as it is right now?

I think we are facing a dilemma:

 - You cannot implicitly convert `T.List[str]` to `T.List[T.Union[str, other]]` or mypy will complain
 - You can implicitly convert `T.Sequence[str]` to `T.Sequence[T.Union[str, other]]`, but as you said, sequences of str's are bad
 - You cannot implicitly convert `ImmutableListProtocol[str]` to `ImmutableListProtocol[T.Union[str, other]]`
 - You could copy the list, but that means a performance hit
 - You could insert a `T.cast` call, but that is ugly and disables many mypy checks.

I think I will be going with the last one.

Volker-Weissmann committed:
Fix accepted types for the input arg of `i18n.merge_file` and `i18n.itstool_join`

Volker-Weissmann commented on mesonbuild/build.py:
```diff
@@ -662,7 +662,7 @@ def get_option(self, key: 'OptionKey') -> T.Union[str, int, bool, 'WrapMode']:
         # TODO: if it's possible to annotate get_option or validate_option_value
         # in the future we might be able to remove the cast here
-        return T.cast('T.Union[str, int, bool, WrapMode]', self.options[key].value)
+        return T.cast(T.Union[str, int, bool, 'WrapMode'], self.options[key].value)
```
Fixed

Volker-Weissmann committed:
Fix all mypy warnings in `mesonbuild/build.py`

Volker-Weissmann committed:
Add a testcase

Feed a custom_target output to cc.preprocess,
then feed that output to a generator.

Volker-Weissmann committed:
Fix error message when attempting to link to a preprocessed source file

Volker-Weissmann commented:
I fixed everything you told me to and made the test pass.

Volker-Weissmann commented on mesonbuild/build.py:
```diff
@@ -2741,6 +2741,9 @@ def _add_output(self, f: File) -> None:
         self.sources_map[f] = o
 
+    def is_linkable_output(self, output: str) -> bool:
+        return False
```
Are you just complaining or is this a call to action?

tristan957 commented:
@dcbaker does your PR make mypy happy for build.py?

dcbaker commented:
@tristan957 no, but it does fix a lot of mypy issues that aren't apparent because it's not 100% clear what's coming in from the interpreter. There's a lot of stuff in there that's pretty broken.

bruchar1 commented on mesonbuild/build.py:
```diff
@@ -808,7 +808,7 @@ def process_objectlist(self, objects):
                 raise InvalidArguments(f'Bad object of type {type(s).__name__!r} in target {self.name!r}.')
 
-    def process_sourcelist(self, sources: T.List['SourceOutputs']) -> None:
+    def process_sourcelist(self, sources: ImmutableListProtocol['SourceOutputs']) -> None:
```
```suggestion
    def process_sourcelist(self, sources: ImmutableListProtocol[SourceOutputs]) -> None:
```

bruchar1 commented on mesonbuild/build.py:
```diff
@@ -1445,7 +1451,7 @@ def link_whole(self, target):
 
     def extract_all_objects_recurse(self) -> T.List[T.Union[str, 'ExtractedObjects']]:
-        objs = [self.extract_all_objects()]
+        objs: T.List[T.Union[str, 'ExtractedObjects']] = [self.extract_all_objects()]
```
```suggestion
        objs: T.List[T.Union[str, ExtractedObjects]] = [self.extract_all_objects()]
```

bruchar1 commented on mesonbuild/dependencies/base.py:
```diff
@@ -109,7 +109,7 @@ def __init__(self, type_name: DependencyTypeName, kwargs: T.Dict[str, T.Any]) ->
         # If None, self.link_args will be used
         self.raw_link_args: T.Optional[T.List[str]] = None
-        self.sources: T.List[T.Union['FileOrString', 'CustomTarget', 'StructuredSources']] = []
+        self.sources: T.List[T.Union['File', 'CustomTarget', 'StructuredSources']] = []
```
```suggestion
        self.sources: T.List[T.Union[File, CustomTarget, StructuredSources]] = []
```

bruchar1 commented on mesonbuild/modules/__init__.py:
```diff
@@ -41,6 +42,9 @@ class ModuleState:
     needed so modules does not touch any other part of Meson internal APIs.
     """
 
+    environment: 'Environment'
+    subdir: 'str'
```
```suggestion
    environment: Environment
    subdir: str
```

bruchar1 commented on mesonbuild/modules/i18n.py:
```diff
@@ -246,7 +249,7 @@ def gettext(self, state: 'ModuleState', args: T.Tuple[str], kwargs: 'Gettext') -
         extra_args = kwargs['args']
         targets: T.List['Target'] = []
-        gmotargets: T.List['build.CustomTarget'] = []
+        gmotargets: T.List['build.Target'] = []
```
```suggestion
        gmotargets: T.List[build.Target] = []
```

bruchar1 reviewed (commented):
With `annotations` from `__future__`, you shouldn't need to quote types. I know many types are quoted for historical reasons, but if we touch them, I think it's better to correct them.
